source/clast.c

```diff
@@ -2110,3 +2114,463 @@ void clast_filter(struct clast_stmt *node,
+        /* unroll_jam_loop_body(next_stmt, ufactor, iterator); */
+    } else if (CLAST_STMT_IS_A(s, stmt_for)) {
+        loop = (struct clast_for *)s;
+        unroll_jam_loop_body(loop->body, ufactor, iterator, stride);
+        unroll_jam_loop_body(s->next, ufactor, iterator, stride);
+    } else if (CLAST_STMT_IS_A (s, stmt_guard)){
+        g = (struct clast_guard *)s;
+        unroll_jam_loop_body(g->then, ufactor, iterator, stride);
+        unroll_jam_loop_body(s->next, ufactor, iterator, stride);
+    }
+    return;
+}
+
+/* Updates the loop upper bound during unroll jam. The new upper bound is
+ * given by (UB - (unroll factor-1)* loop stride) where UB is the upper bound
+ * of the original loop. */
+void update_loop_upper_bound (struct clast_for *loop) {
+    cloog_int_t i, j;
+    struct clast_expr *new_term1, *new_term2;
+    struct clast_reduction *new_ub;
+
+    cloog_int_init(i);
+    cloog_int_init(j);
+    cloog_int_set_si(i,1);
+    cloog_int_sub(j, loop->stride, i);
+    cloog_int_neg(j,j);
+    new_term1 = &(new_clast_term(j, NULL))->expr;
+    new_term2 = &(new_clast_term(i, loop->UB))->expr;
+    new_ub = new_clast_reduction(clast_red_sum, 2);
+    new_ub->elts[0] = new_term2;
+    new_ub->elts[1] = new_term1;
+    loop->UB = &new_ub->expr;
+    cloog_int_clear(i);
+    cloog_int_clear(j);
+}
+
+/* Unroll jam a loop given by the ast node s. */
+void unroll_jam_loop(struct clast_stmt *s) {
+    struct clast_for *loop;
+    assert (CLAST_STMT_IS_A(s, stmt_for));
+    loop = (struct clast_for *)s;
+    unroll_jam_loop_body(loop->body, loop->ufactor, loop->iterator,
+            loop->stride);
+    update_loop_stride(loop);
+    update_loop_upper_bound(loop);
+}
+
+/* Recursively checks whether an expression e containts the string given by
+ * iterator. Returns true if iterator is present. Else returns false. */
+int does_expr_contain(struct clast_expr *e, const char* iterator) {
+    struct clast_name *n;
+    struct clast_term *t;
+    struct clast_binary *b;
+    struct clast_reduction *r;
+    int i;
+    if (!e)
+        return 0;
+    if (e->type == clast_expr_name) {
+        n = (struct clast_name *)e;
+        return (strcmp(n->name, iterator)?0:1);
+    }
+    if (e->type == clast_expr_term) {
+        t = (struct clast_term *)e;
+        return does_expr_contain(t->var, iterator);
+    }
+    if (e->type == clast_expr_bin) {
+        b = (struct clast_binary *)e;
+        return does_expr_contain(b->LHS, iterator);
+    }
+    /* This has to be a reduction type. */
+    assert (e->type == clast_expr_red);
+    r = (struct clast_reduction *)e;
+    for (i=0; i<r->n; i++) {
+        if (does_expr_contain(r->elts[i], iterator))
+            return 1;
+    }
+    return 0;
+}
+
+/* Checks whether the LHS or RHS of on equation containts the string given
+ * by iterator. Returns 1 if the string is present else returns 0. */
+int does_equation_contain(struct clast_equation *e, const char *iterator) {
```

Comment:
Use static for local methods.

Reply:
Done. All local methods are now declared as static